contracts/InterestRates/InterestRateDebtToken.sol

```diff
@@ -0,0 +1,96 @@
+import { ERC20, ERC20Capped } from "../common/ERC20Capped.sol";
+import { ERC20Indexable } from "../ERC20Indexable.sol";
+import { PositionManagerDependent } from "../PositionManagerDependent.sol";
+import { IInterestRatePositionManager } from "./IInterestRatePositionManager.sol";
+
+contract InterestRateDebtToken is ERC20Indexable {
+    // --- Types ---
+
+    using Fixed256x18 for uint256;
+
+    // --- Events ---
+
+    event IndexIncreasePerSecondSet(uint256 indexIncreasePerSecond);
+
+    // --- Immutables ---
+
+    IERC20 immutable collateralToken;
+
+    // --- Variables ---
+
+    uint256 internal storedIndexUpdatedAt;
+
+    uint256 public indexIncreasePerSecond;
+
+    // --- Constructor ---
+
+    constructor(
+        address positionManager_,
+        string memory name_,
+        string memory symbol_,
+        IERC20 collateralToken_,
+        uint256 cap_,
+        uint256 indexIncreasePerSecond_
+    )
+        ERC20Indexable(positionManager_, name_, symbol_, cap_)
+    {
+        storedIndexUpdatedAt = block.timestamp;
+        collateralToken = collateralToken_;
+        setIndexIncreasePerSecond(indexIncreasePerSecond_);
+    }
+
+    // --- Functions ---
+
+    function mint(address to, uint256 amount) public virtual override {
+        updateIndexAndPayFees();
+        super.mint(to, amount);
+    }
+
+    function burn(address from, uint256 amount) public virtual override {
+        updateIndexAndPayFees();
+        super.burn(from, amount);
+    }
+
+    function currentIndex() public view virtual override returns (uint256) {
+        return storedIndex.mulUp(INDEX_PRECISION + indexIncreasePerSecond * (block.timestamp - storedIndexUpdatedAt));
+    }
+
+    function updateIndexAndPayFees() public {
+        uint256 currentIndex_ = currentIndex();
+        _payFees(currentIndex_);
+        storedIndexUpdatedAt = block.timestamp;
+        storedIndex = currentIndex_;
+        emit IndexUpdated(currentIndex_);
+    }
+
+    function setIndex(uint256 backingAmount) public virtual override {
+        _payFees(currentIndex());
+        storedIndexUpdatedAt = block.timestamp;
+        super.setIndex(backingAmount);
+    }
+
+    function setIndexIncreasePerSecond(uint256 indexIncreasePerSecond_) public onlyOwner {
+        indexIncreasePerSecond = indexIncreasePerSecond_;
+        emit IndexIncreasePerSecondSet(indexIncreasePerSecond_);
+    }
+
+    function unpaidFees() external view returns (uint256) {
+        return _unpaidFees(currentIndex());
+    }
+
+    function _unpaidFees(uint256 currentIndex_) private view returns (uint256) {
+        return totalSupply().mulDown(currentIndex_ - storedIndex);
+    }
+
+    function _payFees(uint256 currentIndex_) private {
+        uint256 unpaidFees = _unpaidFees(currentIndex_);
+        if (unpaidFees > 0) {
+            IInterestRatePositionManager(positionManager).mintFees(collateralToken, unpaidFees);
+        }
```

Comment:
Once cap is reached, it wont be possible to adjust or liquidate positions due to interest accrued, no ? 
Only option here would be to increase the cap, but soon that wont be enough due to continious interest

Reply:
NVM
Since, the cap checks `ERC20.totalSupply()` instead of overridden total supply from indexable, the above case would not occur

```solidity
 function _mint(address account, uint256 amount) internal virtual override {
        if (ERC20.totalSupply() + amount > cap) {
            revert ERC20ExceededCap();
        }
        super._mint(account, amount);
    }
 ```

Reply:
Very good point. We will need to think a little bit about it. Probably it makes sense to check ERC20.totalSupply() because that doesn't count interest accrued. However would probably require some accounting to increase caps based on current index over some periods of time.

Reply:
On top of this, Debt token is not minted here, only R tokens, so we should be good here.

Reply:
https://github.com/raft-fi/contracts/pull/455

Reply:
hmm nice, good catch, lgtm

Reply:
Hmm I think this new update https://github.com/raft-fi/contracts/pull/455 actually made the issue I was talking about possible 
Yeah it is true that only R tokens are minted and debt tokens are not 
however each time `payfees` would be called it would increase the debt, thereby index and thereby totalsupply().
hence once the cap is reached, it won't be possible to take any action.
since protocol would try to pay unpaid fees first

Reply:
makes sense, nice, lgtm :)